source/fe/fe_hermite.cc

```diff
@@ -0,0 +1,840 @@
+FE_Hermite<dim, spacedim>::FE_Hermite(const unsigned int reg)
+  : FE_Poly<dim, spacedim>(
+      internal::get_hermite_polynomials<dim>(reg),
+      FiniteElementData<dim>(internal::get_hermite_dpo_vector(dim, reg),
+                             1,
+                             2 * reg + 1,
+                             (reg ? FiniteElementData<dim>::H2 :
+                                    FiniteElementData<dim>::H1)),
+      std::vector<bool>(Utilities::pow(2 * (reg + 1), dim), false),
+      std::vector<ComponentMask>(Utilities::pow(2 * (reg + 1), dim),
+                                 std::vector<bool>(1, true)))
+  , regularity(reg)
+{}
+
+
+
+template <int dim, int spacedim>
+std::string
+FE_Hermite<dim, spacedim>::get_name() const
+{
+  std::ostringstream name_buffer;
+  name_buffer << "FE_Hermite<" << dim << "," << spacedim << ">("
+              << this->regularity << ")";
+  return name_buffer.str();
+}
+
+
+
+template <int dim, int spacedim>
+std::unique_ptr<FiniteElement<dim, spacedim>>
+FE_Hermite<dim, spacedim>::clone() const
+{
+  return std::make_unique<FE_Hermite<dim, spacedim>>(*this);
+}
+
+
+
+template <int dim, int spacedim>
+std::vector<unsigned int>
+FE_Hermite<dim, spacedim>::get_lexicographic_to_hierarchic_numbering() const
+{
+  return internal::hermite_lexicographic_to_hierarchic_numbering<dim>(
+    this->regularity);
+}
+
+
+
+/*
+ * A special version of this function is needed for Hermite elements,
+ * due to the need to rescale different shape functions according to
+ * the derivative order they represent.
+ */
+template <int dim, int spacedim>
+void
+FE_Hermite<dim, spacedim>::fill_fe_values(
+  const typename Triangulation<dim, spacedim>::cell_iterator &,
+  const CellSimilarity::Similarity cell_similarity,
+  const Quadrature<dim> & /*quadrature*/,
+  const Mapping<dim, spacedim> &                           mapping,
+  const typename Mapping<dim, spacedim>::InternalDataBase &mapping_internal,
+  const dealii::internal::FEValuesImplementation::MappingRelatedData<dim,
+                                                                     spacedim>
+    & /*mapping_data*/,
+  const typename FiniteElement<dim, spacedim>::InternalDataBase &fe_internal,
+  dealii::internal::FEValuesImplementation::FiniteElementRelatedData<dim,
+                                                                     spacedim>
+    &output_data) const
+{
+  // Convert data object to internal data for this class.
+  // Fails with an exception if that is not possible.
+  Assert(
+    (dynamic_cast<const typename FE_Hermite<dim, spacedim>::InternalData *>(
+       &fe_internal) != nullptr),
+    ExcInternalError());
+  const typename FE_Hermite<dim, spacedim>::InternalData &fe_data =
+    static_cast<const typename FE_Hermite<dim, spacedim>::InternalData &>(
+      fe_internal);
+
+  Assert(
+    (dynamic_cast<const typename MappingHermite<dim, spacedim>::InternalData *>(
+       &mapping_internal) != nullptr) ||

```

Comment:
Let's keep this simpler and not think about a mapping class that might be implemented in the future.